chain/index/api_test.go

```diff
@@ -0,0 +1,483 @@
+package index
+
+import (
+	"context"
+	pseudo "math/rand"
+	"testing"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/stretchr/testify/require"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+func TestValidateIsNullRound(t *testing.T) {
+	ctx := context.Background()
+	rng := pseudo.New(pseudo.NewSource(time.Now().UnixNano()))
+	headHeight := abi.ChainEpoch(100)
+
+	tests := []struct {
+		name           string
+		epoch          abi.ChainEpoch
+		setupFunc      func(*SqliteIndexer)
+		expectedResult bool
+		expectError    bool
+		errorContains  string
+	}{
+		{
+			name:           "happy path - null round",
+			epoch:          50,
+			expectedResult: true,
+		},
+		{
+			name:  "failure - non-null round",
+			epoch: 50,
+			setupFunc: func(si *SqliteIndexer) {
+				insertTipsetMessage(t, si, tipsetMessage{
+					tipsetKeyCid: randomCid(t, rng).Bytes(),
+					height:       50,
+					reverted:     false,
+					messageCid:   randomCid(t, rng).Bytes(),
+					messageIndex: 0,
+				})
+			},
+			expectError:   true,
+			errorContains: "index corruption",
+		},
+		{
+			name:           "edge case - epoch 0",
+			epoch:          0,
+			expectedResult: true,
+		},
+		{
+			name:           "edge case - epoch above head",
+			epoch:          headHeight + 1,
+			expectedResult: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			si, _, _ := setupWithHeadIndexed(t, headHeight, rng)
+			defer func() { _ = si.Close() }()
+
+			if tt.setupFunc != nil {
+				tt.setupFunc(si)
+			}
+
+			res, err := si.validateIsNullRound(ctx, tt.epoch)
+
+			if tt.expectError {
+				require.Error(t, err)
+				if tt.errorContains != "" {
+					require.Contains(t, err.Error(), tt.errorContains)
+				}
+			} else {
+				require.NoError(t, err)
+				require.NotNil(t, res)
+				require.Equal(t, tt.expectedResult, res.IsNullRound)
+				require.Equal(t, uint64(tt.epoch), res.Height)
+			}
+		})
+	}
+}
+
+func TestFailureHeadHeight(t *testing.T) {
+	ctx := context.Background()
+	rng := pseudo.New(pseudo.NewSource(time.Now().UnixNano()))
+	headHeight := abi.ChainEpoch(100)
+
+	si, head, _ := setupWithHeadIndexed(t, headHeight, rng)
+	defer func() { _ = si.Close() }()
```

Comment:
I think `t.Cleanup()` is the preferred way to do these in tests rather than using defer

Reply:
Done in ALL tests.